### DISCUSSION TITLE
Do you have a migration guide from @polkadot/api?

### DISCUSSION THREAD

**Discussion by decentration** [polkadot-api/polkadot-api#888] [Q&A] [answered]
Hi there friends, we will soon be starting our transition from polkadot/api to polkadot-api "PAPI".

Our project uses a whole range of the @polkadot/api library. 

1.  Do you only replace all (or most of?) the features from @polkadot/api? or are there other @polkadot/x libraries we should also replace?

we use this:

```
//...
    "@polkadot/api": "^12.2.1",
    "@polkadot/api-augment": "^12.2.1",
    "@polkadot/api-base": "^12.4.2",
    "@polkadot/api-contract": "^12.2.1",
    "@polkadot/api-derive": "^12.4.2",
    "@polkadot/extension-dapp": "^0.47.3",
    "@polkadot/extension-inject": "^0.47.3",
    "@polkadot/hw-ledger": "^13.0.2",
    "@polkadot/keyring": "^13.0.2",
    "@polkadot/networks": "^13.0.2",
    "@polkadot/phishing": "^0.23.1",
    "@polkadot/rpc-augment": "^12.2.1",
    "@polkadot/rpc-core": "^12.2.1",
    "@polkadot/rpc-provider": "^12.2.1",
    "@polkadot/types": "^12.2.1",
    "@polkadot/types-augment": "^12.2.1",
    "@polkadot/types-codec": "^12.2.1",
    "@polkadot/types-create": "^12.2.1",
    "@polkadot/types-known": "^12.2.1",
    "@polkadot/types-support": "^12.2.1",
    "@polkadot/ui-keyring": "^3.1.4",
    "@polkadot/util": "^12.6.2",
    "@polkadot/util-crypto": "^12.6.2",
    "@polkadot/wasm-crypto": "^7.3.2",
    //...
```
    
2. assuming polkadot-api is aiming to be the more contemporary version do you have a migration guide? so we can prioritise this transition as smoothly as possible. 

An example of changes required:

```
// get the value for an account
const accountInfo = await dotApi.query.System.Account.getValue(
  "16JGzEsi8gcySKjpmxHVrkLTHdFHodRepEz8n244gNZpr9J",
)
```

in polkadot/api this was:

`const accountInfo = await api.query.system.account(...)...`

- So are all pallets and methods and parameters no longer lower case?
- is getValue also needed to query, rather than just passing the value itself?

This is just one example of the changes required for migration. 

I will do my best to create a guide as breadcrumbs for other projects. 

But any support, guidance would be appreciated. 

Cheers! ✌️

**Comment by josepot** [comment, marked as answer]
> Do you have a migration guide from @polkadot/api?

No, we don’t have a migration guide - and for good reason. Creating one wouldn’t really work. Here’s why:  

PolkadotJS has a very "leaky" design, with many libraries tightly coupled to each other and without clear boundaries. Polkadot-API focuses on creating better, simpler, and more modular interfaces.  

If you're trying to migrate from PJS to `polkadot-api`, it’s less about "mapping" one approach to the other and more about rethinking how your app interacts with the API. This is like moving from JQuery to a modern framework like SolidJS or Svelte - it's not about step-by-step migration but about embracing a new way of working. Would a migration guide from JQuery to Solid or Svelte make sense? Probably not, right? Well, the same applies here.

Although, @Tbaut gave a pretty good talk at Sub0 sharing how you can migrate a DApp from PJS to PAPI, it's probably worth [checking it out](https://www.youtube.com/watch?v=fRI96YD105E).

Also, It’s crucial to understand that the `polkadot-api` library is specifically designed to help **DApp developers** build decentralized applications efficiently. However, if you’re working on something other than a DApp—like developer tools, meta-programming frameworks, or no-code platforms—you should probably either get familiar with our internal libraries or should should create your own.

About the libraries that you have listed above, these are some thoughts:

- How can you migrate from an extremely leaky and convoluted interface [like the PJS rpc-provider interface](https://github.com/polkadot-js/api/blob/master/packages/rpc-provider/src/types.ts#L51-L68) to a simple, powerful, enhanceable [interface like the one that we use at PAPI](https://github.com/polkadot-api/polkadot-api/blob/main/packages/json-rpc/json-rpc-provider/src/index.ts#L1-L8)? You simply can't. It just doesn't make sense.

- Why would anyone use `@polkadot/wasm-crypto` (or `@polkadot/util-crypto`) these days, when we have libraries like `@noble/hashes`, `@noble/curves` and ? Polkadot-API (unlike PJS) is [not tightly coupled to any particular crypto library](https://papi.how/signers#signers) and it doesn't impose a leaky crypto API.

- Why would anyone use `@polkadot/hw-ledger` when libraries like [`@polkadot-api/ledger-signer`](https://github.com/polkadot-api/polkadot-api/tree/main/packages/signers/ledger-signer) and/or [`@polkadot-api/merkleize-metadata`](https://github.com/polkadot-api/polkadot-api/tree/main/packages/merkleize-metadata) exist? 🤷 (Even the PJS extension uses the [`@polkadot-api/merkleize-metadata`](https://forum.polkadot.network/t/polkadot-js-new-ledger-support-completion/9559)). It really makes no sense to me...

- If you get into polkadot-api, then you will probably realize very soon that libraries like `@polkadot/api-augment`, `@polkadot/api-base` and even `@polkadot/networks` don't really make that much sense anymore...
